### ISSUE TITLE
sidekiq_alive causes Redis CPU to max out on big deployments

### ISSUE THREAD

**Issue by Fire-Dragon-DoL** [arturictus/sidekiq_alive#69]
Hi!
We recently performed an upgrade from v2.1.4 to v2.1.5 that caused a system outage: on boot, Sidekiq maxed out Redis CPU. 
The machine where Redis is hosted usually doesn't see more than 10% CPU usage.

The investigation for the causes is still in progress, the code that is potentially causing the issue was [contributed by our company to sidekiq_alive](https://github.com/arturictus/sidekiq_alive/compare/v2.1.4...v2.1.5), so we caused our own problem, but it's important to consider that other big deployments (million jobs) might face the same issue, the version increase should probably be at least MINOR, not PATCH, given the kind of outage it caused.

I will be investigating in how this happened, so I will be able to provide more details.

**Comment by arturictus** [comment]
@Fire-Dragon-DoL thanks for letting me know.
Any news?

**Comment by Fire-Dragon-DoL** [comment]
> @Fire-Dragon-DoL thanks for letting me know. Any news?

Hi @arturictus , I don't have "direct" news, the investigation is still going, we are waiting for a replica of production Redis with data scrambled to replicate the behavior.
I gathered some numbers and the software I work on processes _at least_ 10M jobs per day, which suggests the `scan` with 1000 count might be the problem, but this is still an hypothesis.

I'll keep you posted, the ticket is ongoing.

**Comment by Fire-Dragon-DoL** [comment]
@arturictus I was able to track what was going on!

The problem is the following method:

```ruby
    def registered_instances
      redis.match("#{config.registered_instance_key}::*")
    end
```

This method is called at:

```ruby
    def successful_startup_text
      "Successfully started sidekiq-alive, registered instances: #{registered_instances.join("\n\s\s- ")}"
    end
```

Which happens synchronously, at boot. Now, `scan` is not synchronous, but called in a tight ruby loop will behave _similarly_ to `keys *`, essentially iterating over all the keys of the database.
We use a sidekiq extension that write locks as keys in Redis, so there are more than 3M keys in this deployment.

Writing that message causes the boot time to increase from a ~30 seconds to over 5 minutes.

Changing the approach could be helpful in this context.

I noticed also this method that could be problematic: https://github.com/arturictus/sidekiq_alive/blob/be831f54efccb675c2ea5fd1bb7245a970df965e/lib/sidekiq_alive/worker.rb#L18-L22

```ruby
    def hostname_registered?(hostname)
      SidekiqAlive.registered_instances.any? do |ri|
        /#{hostname}/ =~ ri
      end
    end
```

However it seems to be used only in tests and as such, not affecting anything.

The fast solution is to inform that boot has happened without listing those keys. The "maybe right" solution would be to use a single key instead of a group of keys.

`SCAN` with a pattern is still `O(n)` operation, the filtering happens _after the results are fetched_, before being returned to the client.
